Per #590: Handle keywords in nested maps in configuration

fbergroth opened the pull request:
* eglot.el (eglot--dump-configuration): Add function
(eglot-signal-didChangeConfiguration): Dump configuration recursively

fbergroth commented:
@joaotavora I guess this works if elisp lists are used for json maps, and elisp vectors are used for lists. Ugh.

fbergroth commented:
If we don't touch the configuration at all, the user would be free to write either
```
'((key1 . "value1") (key2 . "value2")))
```
or
```
'(:key1 "value1" :key2 "value2"))
```
which both encodes as 
```
{"key1":"value1","key2":"value2"}
```

fbergroth commented:
I think the sensible thing is to not touch the configuration, and refer to the documentation of `json-serialize` in the docstring of `eglot-view-configuration`.

I also added the command `eglot-view-configuration` which could be helpful.

joaotavora commented:
In a now-deleted comment you wrote:

> @joaotavora I guess this works if elisp lists are used for json maps, and elisp vectors are used for lists. Ugh.

Personally I don't think it's _that_ Ugh, as long as it is cleanly documented somehwere.  In Eglot we do use plists for JSON objects and elisp sequences for JSON arrays.  We could choose other methods, but I happened to choose this one which I think is the less verbose one (prints better, too, certainly better than the hash-table equivalent).

Does you PR change any of this, or does it mostly fix the bugs and improve the documenation?

fbergroth commented:
Yeah, I later deleted that comment since I came to realize it wasn't that Ugh.

I pushed some new changes.

The first commit are for the variant which recursively (instead of just at the top level) rewrites `((key . value))` as `(:key value)`, which works, but is confusing and unnecessary, so I reverted it.

I think just keeping the value untouched and referring to `json-serialize` will make it easier for users. It may create friction since it can break some configurations which is unfortunate.

I've added a helper `eglot-show-configuration` which dumps the configuration value as json.

Have a look and feel free to change anything.

fbergroth commented:
Ah, I missed that `json-encode` is used if `json-serialize` isn't available, which encodes a bit differently. 

`(json-serialize '(a 123))` encodes as `{"a": 123}`, while `(json-encode '(a 123))` encodes as `["a", 123]`.

Hmm, so perhaps the recursive rewrite should be kept?

joaotavora commented:
> (json-serialize '(a 123)) encodes as {"a": 123}, while (json-encode '(a 123)) encodes as ["a", 123].

IIRC the behavior of both these serializers depends very much on additional arguments, in the first case, and special variables in the second case. It should be possible to make these two functions behave equivalently.

fbergroth commented:
@joaotavora Sorry, I got a bit confused yesterday.

Now I went with a run-time check of the structure as you initially suggested. I think it looks better now.

fbergroth commented:
@skangas please have a look as well.

The configuration you used here https://github.com/joaotavora/eglot/issues/590#issuecomment-1008533473 previously encoded as `{'pylsp': {':plugins': {':pydocstyle' ...`, and the settings was ignored by the language server. With this PR, the dictionary keys won't have the colons.

fbergroth commented:
Ah, actually it does work with `json-encode` but not with `json-serialize`. Now it should work with both though.

joaotavora commented:
Yeah, if this is a self-contained improvement that it doesn't make sense to split, please to squash as usual. Make history much nicer to read.  OTOH if some early commit is laying groundwork for multiple possible other commits, it makes sense to isolate it.

fbergroth commented:
The commits are a bit sloppy. I didn't want to spend too much time on this before getting some feedback from you. I can clean it up once we agree on the diffs.

Ideally the test should include `t nil :json-false` and call `jsonrpc--json-encode` twice. Once to trigger `json-serialize`, and once with `json-serialize` uninterned, so `json-encode` is used.

joaotavora commented:
> The commits are a bit sloppy. I didn't want to spend too much time on this before getting some feedback from you. I can clean it up once we agree on the diffs.

Makes total sense 👍 to me.  Will try to find time soon.

skangas commented on eglot.el:
```diff
@@ -2002,19 +2002,29 @@ that can be converted to JSON.")
+   ((listp value)
+    (cl-loop for elem in value
+             unless (consp elem) do
+             (user-error "invalid alist entry `%S' (expected cons)" elem)
```
- Probably better as `"Invalid alist entry in configuration: %S"`
- Shouldn't this be just an `error`?

fbergroth replied:
> * Probably better as `"Invalid alist entry in configuration: %S"`

Sure! Perhaps the string should even be prefixed with `eglot: `? 

> * Shouldn't this be just an `error`?

It'll be signalled when the `eglot-workspace-configuration` is invalid, such as when the user is using a plist instead of an alist. I think `user-error` is appropriate, can you elaborate?

skangas replied:
How about just using `eglot--error`? I think then we get the prefix for free.

AFAIK, the only reason for using `user-error` is to not get a stacktrace when `debug-on-error` is t. However, I think it would be nice to get a stacktrace in this case.

fbergroth replied:
Makes sense, I wasn't aware of that :+1:

skangas commented on eglot.el:
```diff
@@ -2002,19 +2002,29 @@ that can be converted to JSON.")
+             collect (let ((k (car elem)))
+                       (if (keywordp k) k (intern (format ":%s" k))))
+             collect (eglot--check-conf (cdr elem))))
+   (t (user-error "invalid configuration value `%S'" value))))
```
- Probably better as `"Invalid configuration value: %S"`
- Shouldn't this be just an `error`?

skangas commented on README.md:
```diff
@@ -157,10 +157,10 @@ a section for `go-mode`, the file's contents become:
  (go-mode
   . ((eglot-workspace-configuration
-      . ((:gopls . (:usePlaceholders t)))))))
+      . ((:gopls (:usePlaceholders . t)))))))
```
How about adding the examples from the README as tests?

fbergroth replied:
Sure. In some form, either in the big test or as separate ones.

skangas replied:
Sounds good.

skangas commented:
Could we also have a test for the case you mentioned in #590: `'((:a (:b . t)))`?

fbergroth commented on eglot-tests.el:
```diff
@@ -1162,6 +1162,13 @@ are bound to the useful return values of
+  (let ((value (eglot--check-conf '((:k1 (:k2 . [a "b" 1 2.0]) (k3 . v1)))))
+        (expected (string-replace "'" "\"" "{'k1':{'k2':['a','b',1,2.0],'k3':'v1'}}")))
+    (should (equal (json-encode value) expected))
+    (should (equal (json-serialize value) expected))))
```
The important test is validating that whatever goes through `eglot--check-conf` without error should encode to the same value regardless of  `json-serialize` being available or not. I also wrote:

> Ideally the test should include `t nil :json-false` and call `jsonrpc--json-encode` twice. Once to trigger `json-serialize`, and once with `json-serialize` uninterned, so `json-encode` is used.

... but it's a bit awkward.  `eglot--check-conf` may perhaps better belong in `jsonrpc` so that internals aren't leaked, but I haven't looked into it.

fbergroth replied:
`eglot--check-conf` can perhaps be called `jsonrpc-safe-encode-transform`

skangas replied:
I feel like I'm missing something: why should `jsonrpc.el` have code that relates to how `eglot` represents it's configuration. Does that really make sense?

BTW, we should probably add an `fboundp` somewhere, in case `json-serialize` is not available.

fbergroth replied:
> I feel like I'm missing something: why should `jsonrpc.el` have code that relates to how `eglot` represents it's configuration. Does that really make sense?

`eglot--check-conf` is really checking that a particular value is structured in a way so that it encodes the same regardless of `json-serialize` being available. It's should allow what's common between `json-encode` and `json-serialize` (or transform the value such that it becomes that). It could make sense to send any user input through it, but probably it makes more sense to keep it as is. 
 
> BTW, we should probably add an `fboundp` somewhere, in case `json-serialize` is not available.

There is already `(skip-unless (fboundp 'json-serialize))`.

fbergroth replied:
Another point why it could belong in jsonrpc, is that the chosen representation is really based on implementation details of jsonrpc, which eglot should not need to know about. What do you think @joaotavora?

fbergroth commented on eglot-tests.el:
```diff
@@ -1162,6 +1162,13 @@ are bound to the useful return values of
+(ert-deftest eglot--compare-json-serializers ()
+  (skip-unless (fboundp 'json-serialize))
+  (let ((value (eglot--check-conf '((:k1 (:k2 . [a "b" 1 2.0]) (k3 . v1)))))
+        (expected (string-replace "'" "\"" "{'k1':{'k2':['a','b',1,2.0],'k3':'v1'}}")))
```
I think stuffing all the cases in a single example is fine (perhaps the existing one can be made to look a bit better), but I can change if you prefer a bunch of smaller ones.

skangas replied:
I generally prefer several smaller ones, even if it means some repetition: IME, it is then easier to add new cases, and it is easier to compare the various cases to the various branches.

But yes, it is partly a matter of style, so I don't want to insist too strongly on it. Maybe @joaotavora has an opinion.

doolio commented:
@fbergroth was there no need to have this merged in the end?

fbergroth commented:
I think the changes in eglot.el are useful, but João never got back to this, and I didn't want to pick it up after several months.

joaotavora commented:
Whoops, sorry indeed that's my bad.  The problem is that all the discussion is out of my mental cache now. But on a brief overview, it does look useful and sane.  

@doolio can you perhaps state your need/use case for this?  What caught your attention in this PR?

doolio commented:
Thank you both for the quick responses!

I came across this PR when searching through the repository for any mention of `eglot-workspace-configuration`. I'm trying to configure multiple server settings in a `.dir-locals.el` file as recommended by the README but where the example is when setting only a single server option. It does though provide an example of setting a single option on more than one server. [Aside: Can one use multiple servers for the same language or is that just silly. I'm wondering if one can use `python-lsp-server` and `pyright` at the same time as I believe they provide only some common features but perhaps where they do provide the same they would conflict.] I was struggling to wrap my head around `alist`s and `plist`s. I've been using Emacs for a few years and am trying to learn to programme but have not yet really written any Emacs Lisp. 

It seems across various issues, PRs and discussions that users have struggled to set this variable correctly. I find several variations and attempts to do so. I thought this PR was an effort to simplify the setting of this variable for users. Is that correct?

I believe to configure that variable correctly we need to add a _plist_ as the `eglot-workspace-configuration` _alist_ `VALUE` value within the `.dir-locals.el` nested _alists_ if I understood the README correctly. But then this _plist_ can contain nested _plists_. Please correct me if I'm mistaken. 

It seems one can use symbols or keyword symbols (i.e. :symbol) for the `eglot-workspace-configuration` _alist_ `SECTION` key  but also for the property name (or key) in the _plist_. It also seems from [this Emacs manual page](https://www.gnu.org/software/emacs/manual/html_node/elisp/Parsing-JSON.html) that the _plist_ values for the JSON `false` and `null` keywords are by default the keyword symbols `:false` and `:null` respectively. But it seems at least such keyword symbol values are not accepted by the `python-lsp-server`. I've tried variations such as `':false`, `'false` etc. but to no avail. I've tried using `nil` for `false` and it seems to work despite the same Emacs manual page stating `nil` represents {}, the empty JSON object - but maybe that is why it works? Do you know what symbol we should use to represent the JSON array `[]` which is another default value of the `python-lsp-server`? I appreciate the values accepted by the servers are nothing to do with `eglot` and that they likely differ between servers.

My attempt so far to configure the `python-lsp-server` as follows works. It enables `flake8` as a checker disabling the other checkers which flake8 wraps around and thus avoids duplicate reports in `flymake`. One must also set the Emacs variable `python-flymake-command` to ("flake8" "-") as well and obviously have the relevant python packages installed. I've disabled the `autopep8` and `yapf` formatters as I use black via the `python-lsp-black` plugin for the server.

```
;;; Directory Local Variables
;;; For more information see (info "(emacs) Directory Variables")

((python-mode
  . ((eglot-workspace-configuration
      . ((:pylsp
          . (:configurationSources ["flake8"]
                                   :plugins (:autopep8 (:enabled nil)
                                                       :flake8 (:enabled t)
                                                       :mccabe (:enabled nil)
                                                       :pycodestyle (:enabled nil)
                                                       :pydocstyle (:enabled nil :convention "pep257")
                                                       :pyflakes (:enabled nil)
                                                       :yapf (:enabled nil)))))))))
```
Is using `nil` here wrong? I suspect so but as I stated above none of the different variations I tried worked. 

If there is a wiki page under this project I'd be happy to add a `.dir-locals.el` file for the `python-lsp-server` to include all the possible settings as given [here](https://github.com/python-lsp/python-lsp-server/blob/develop/CONFIGURATION.md). It might encourage other users to do the same for other servers? I'd certainly add any for other servers I try.

Thanks for your time and this package.

astoff commented on eglot.el:
```diff
@@ -2002,19 +2003,44 @@ that can be converted to JSON.")
+    (cl-map 'vector #'eglot--check-conf value))
+   ((listp value)
+    (cl-loop for elem in value
+             unless (consp elem) do
+             (user-error "invalid alist entry `%S' (expected cons)" elem)
```
Where it the case of a plist handled? It seems to me that a plist would always throw this error.

astoff commented on README.md:
```diff
@@ -139,7 +139,7 @@ file named `.dir-locals.el` in the project's root:
 ((python-mode
   . ((eglot-workspace-configuration
-      . ((:pyls . (:plugins (:jedi_completion (:include_params t)))))))))
+      . ((:pyls (:plugins (:jedi_completion (:include_params . t)))))))))
```
Why did you include a dot after `:include_params`? AFAIU, the only place a cons cell like this can appear in the Elisp -> JSON translation is as an entry of an alist, and this cons appears as an value in a plist.

In fact, `(json-encode '(:jedi_completion (:include_params . t)))` gives an error and so does `json-serialize`.

doolio replied:
> (json-encode '(:jedi_completion (:include_params . t)))

Thanks for this snippet. With this without the erroneous dot, it seems `json-serialize` encodes the JSON keywords as expected but `json-encode` only some namely `:true` and `:[]`.

```
(json-encode '(:autopep8 (:enabled t)))         ; "{\"autopep8\":{\"enabled\":true}}"
(json-encode '(:autopep8 (:enabled nil)))       ; "{\"autopep8\":{\"enabled\":null}}"
(json-encode '(:autopep8 (:enabled :null)))     ; "{\"autopep8\":{\"enabled\":\"null\"}}"
(json-encode '(:autopep8 (:enabled :false)))    ; "{\"autopep8\":{\"enabled\":\"false\"}}"
(json-encode '(:autopep8 (:enabled [])))        ; "{\"autopep8\":{\"enabled\":[]}"

(json-serialize '(:autopep8 (:enabled t)))      ; "{\"autopep8\":{\"enabled\":true}}"
(json-serialize '(:autopep8 (:enabled nil)))    ; "{\"autopep8\":{\"enabled\":{}}}"
(json-serialize '(:autopep8 (:enabled :null)))  ; "{\"autopep8\":{\"enabled\":null}}"
(json-serialize '(:autopep8 (:enabled :false))) ; "{\"autopep8\":{\"enabled\":false}}"
(json-serialize '(:autopep8 (:enabled [])))     ; "{\"autopep8\":{\"enabled\":[]}"
```
I believe `jsonrpc` uses `json-encode` as an [alias](https://github.com/emacs-mirror/emacs/blob/191c4fe249e1191b770a6051c0f24c5783ac09af/lisp/jsonrpc.el#L496) to `json-serialize` so I don't quite understand why `:null` and `:false` were not acceptable in my above `'dir-locals.el` example.

astoff replied:
It's because `jsonrpc--json-encode` modifies what Elisp values these functions use for JSON null and false.

To debug your server configuration, you should use `jsonrpc--json-encode` only.

doolio replied:
OK thanks.

joaotavora replied:
@doolio in case you didn't know this, the behaviour of `json-encode` is affected by multiple dynamic variable bindings. 

I think @astoff is also trying to tell you this.

`jsonrpc.el` sets these variables.  I don't know if you're taking this into account in your tests, or what these tests are supposed to mean.  Eglot does not use `json-serialize` or `json-encode` directly, rather `jsonrpc.el`.  

It's `jsonrpc.el` which sometimes uses one and sometimes uses the other.

doolio replied:
Yes, that's clear now. `jsonrpc--json-encode` produces the same encoding as my first block above. With astoff's suggestion I have determined  I need to use `:json-false` to produce `:false`. By that token shouldn't `jsonrpc.el` set `:null-object` to `:json-null` (rather than `nil`) to produce `:null` and have another term for `:empty-object` set to something like `json-empty` to produce `:{}`?

That is, right now it is:
```
;; Emacs Lisp  |  JSON
;; -------------------------------
;; t           | :true i.e. true
;; :json-false | :false i.e. false
;; nil         | :null i.e. null
;; :[]         | :[] i.e. []
```
but I'm thinking it should be:

```
;; Emacs Lisp  |  JSON
;; -------------------------------
;; t           | :true i.e. true
;; :json-false | :false i.e. false
;; :json-null  | :null i.e. null
;; :[]         | :[] i.e. [], an empty array
;; :json-empty | :{} i.e. {}, an empty object
```

astoff replied:
`:[]` is not a keyword, it gets read as the keyword `:` followed by the empty vector.

doolio replied:
> `:[]` is not a keyword, it gets read as the keyword `:` followed by the empty vector.

Yes, I corrected the above tables. Thanks.

doolio replied:
@joaotavora any thoughts on [this comment](https://github.com/joaotavora/eglot/pull/790#discussion_r968444946)?

astoff replied:
This PR is about how the keys of the JSON objects are handled, so this comment seems a bit unrelated to me. But it is perhaps a bit of a problem that the values in the JSON object have to be provided in the way `jsonrcp--json-serialize` expects them, because that function is convenient for the Lisp programmer but not necessarily the most user-friendly.

For the end user, it would maybe be more friendly to allow the following translations:
```
JSON false <-> :false
JSON true <-> t or :true
JSON null <-> nil or :null
JSON {} <-> I don't know, maybe :{}?
```

doolio replied:
> This PR is about how the keys of the JSON objects are handled, so this comment seems a bit unrelated to me.

True, I could open a new issue under `jsonrpc.el` if João feels it is more appropriate. I added it here as the discussion was fresh in people's minds and so seemed appropriate.

The translations I suggest above I based on [this Emacs manual page](https://www.gnu.org/software/emacs/manual/html_node/elisp/Parsing-JSON.html).

astoff commented:
> I thought this PR was an effort to simplify the setting of this variable for users. Is that correct?

@doolio I don't think it simplifies much for the user, or at least it will not change anything for your attempted configuration. But what do you think of [this suggestion](https://github.com/joaotavora/eglot/pull/1033#issuecomment-1243304059).

doolio commented:
I don't think I'm knowledgeable enough to comment on the specific implementation but certainly your description sounds like it would make it easier for the user to configure.

astoff commented:
If you use `json-encode`, everything works as expected in all cases.

If you use `json-serialize`, then everything works as expected except when:
* there's an alist with string keys — this produces an error;
* there's a plist with string keys — this produces an error;
* there's an alist with keywork keys — this produces a JSON object with keys of the form `":something"` instead of `"something"`, which probably counts as an "unexpected' result.

astoff commented:
Okay, this is what we get in the main branch. I'm pasting what appears in the *Eglot events* and what comes out of that from `jsonrpc--json-encode` when `json-serialize` is available.
1. This is okay.
```
((python-mode
  . ((eglot-workspace-configuration
      . ((:pyls . (:plugins (:jedi_completion (:include_params t)))))))))
=*Eglot events*=>
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
          (:settings
           (:pyls
            (:plugins
             (:jedi_completion
              (:include_params t))))))
=jsonrpc--json-encode=>
"{\"settings\":{\"pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}"
```
2. This is okay.
```
((python-mode
  . ((eglot-workspace-configuration
      . ((:pyls . ((plugins . (:jedi_completion (:include_params t))))))))))
=*Eglot events*=>
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
          (:settings
           (:pyls
            ((plugins :jedi_completion
                      (:include_params t))))))
=jsonrpc--json-encode=>
"{\"settings\":{\"pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}"
```
3. This is not okay.
```
((python-mode
  . ((eglot-workspace-configuration
      . ((:pyls . ("plugins" (:jedi_completion (:include_params t)))))))))
=*Eglot events*=>
Error: jsonrpc--json-encode: Wrong type argument: symbolp, "plugins"
```
4. This is not okay (the JSON object contains `":plugins"` as key).
```
((python-mode
  . ((eglot-workspace-configuration
      . ((:pyls . ((:plugins . (:jedi_completion (:include_params t))))))))))
=*Eglot events*=>
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
          (:settings
           (:pyls
            ((:plugins :jedi_completion
                       (:include_params t))))))
=jsonrpc--json-encode=>
"{\"settings\":{\"pyls\":{\":plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}"
```
5. This is not okay.
```
((python-mode
  . ((eglot-workspace-configuration
      . ((:pyls . (("plugins" . (:jedi_completion (:include_params t))))))))))
=*Eglot events*=>
Error: jsonrpc--json-encode: Wrong type argument: symbolp, "plugins"
```

If I apply `json-encode` to these forms appearing in the `*Eglot events*`, only case 4 changes. We get this:

4bis. This is okay
```
((python-mode
  . ((eglot-workspace-configuration
      . ((:pyls . ((:plugins . (:jedi_completion (:include_params t))))))))))
=*Eglot events*=>
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
          (:settings
           (:pyls
            ((:plugins :jedi_completion
                       (:include_params t))))))
=json-encode=>
"{\"settings\":{\"pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}"
```

I hope this helps.

As to what comes out of the current PR code, I'll let someone else test it.

joaotavora commented:
> I hope this helps.

Yes it does, thanks!

fbergroth commented:
Writing the configuration can be quite trial and error. And it may not be obvious how to debug it, or that there is an error at all, since many language servers will happily accept a bad configuration. I think I had pylsp misconfigured myself at some point without realising it. `eglot-show-configuration` would have helped me there.

https://github.com/joaotavora/eglot/pull/1033#issuecomment-1243304059
> To be honest, I'm not a fan of attempting to fix inconsistencies of the two encoders, which AFAICT are pretty mild.

The current cl-loop that rewrites the sections as keywords kind of does it already. But it does not do it recursively, which I think is a bug.

I probably missed supporting alists since I left this in a draft state and went with a more strict set of values, but that or the dotted syntax sugar can be easily added to this as well...

joaotavora commented:
I've had another cursory look at this PR:

* `eglot--check-conf` needs a docstring, so I can understand exactly what it does.
* I'd like this PR to be split in two commits, if possible.  One of the commits is this "configuration diagnostic" functionality that keeps the current syntax, semantics -- including supposed bugs -- unchanged.  The other commit proposes to fix that bug.
* Since documentation is such an important part of this, I'd like to see the README.md section on `eglot-workspace-configuration` rewritten.


Also, I wouldn't necessarily mind if `eglot-workspace-configuration` is completely redesigned in a backward-incompatible way.  Of course, this would necessitate that that variable is obsoleted.  A new variable, called perhaps `eglot-workspace-plist`, with a clear plist-only structure would be preferable, in my opinion.

astoff commented:
@joaotavora Sorry to insist, but doesn't [this idea](https://github.com/joaotavora/eglot/pull/1033#issuecomment-1243304059) make sense to you? It has the curious side effect that it's backwards compatible. (Also, since `json-serialize` accepts alists, I don't think it's worth the trouble to _forbid_ them, but I do think it should be recommended in the docstring to use plists.)

To recap: we would allow both of these forms in the dir-locals.el and they would be equivalent.
```
((python-mode
  . ((eglot-workspace-configuration
      "pyls.plugins.jedi_completion.include_params" t
      "pyls.plugins.something_else" [1 2 3]))))
;; or equivalently
((python-mode
  . ((eglot-workspace-configuration
      . ((:pyls (:plugins (:jedi_completion (:include_params . t))
                           :something_else [1 2 3]))))))
```

joaotavora commented:
I'm afraid not to fond of these unlispy,  repetitive constructs. But it should be very easy to construct a function that converts from that preferred format to the underlying  plist-based one.

astoff commented:
The problem here, arguably, is that these servers expect these horribly deeply nested things. The pedestrian way to write it may not be elegant, but at least it's legible.

Also, as a thought experiment, suppose I have a base configuration that holds for all projects, say `pyls.plugins.jedi_completion.include_params=true` and then I have a special project where I want to add the extra `pyls.plugins.something_else=[1 2 3]`. If you are using the flattened out representation, you can easily say
```
(setq-local eglot-workspace-configuration (append '("pyls.plugins.something_else" [1 2 3])
                                                  eglot-workspace-configuration))
```
If you are using the deeply nested plists, you need to write your own version of Clojure's `assoc-in`, because I don't think such a thing is available out-of-the-box in Elisp.

astoff commented:
> A new variable, called perhaps eglot-workspace-plist, with a clear plist-only structure would be preferable, in my opinion.

If that's what we want to do, then there's no need to obsolete anything. We can just merge the small change of #1033 so that plists work, and add a warning when `eglot--workspace-configuration` is not a plist.

fbergroth commented:
Unless I am mistaken, if the "strange" keys are not top-level, then each of the cases by @astoff in https://github.com/joaotavora/eglot/pull/790#issuecomment-1244018993 serializes differently by `json-encode` and `json-serialize`.

<details>
  <summary>Output:</summary>

```
$ emacs --batch -l test.el                                                                                                                                                16:17:00

"((python-mode (eglot-workspace-configuration (:pyls :plugins (:jedi_completion (:include_params t))))))
json-encode:    {\"python-mode\":{\"eglot-workspace-configuration\":{\"pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}}
json-serialize: {\"python-mode\":{\"eglot-workspace-configuration\":{\":pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}}"

"((python-mode (eglot-workspace-configuration (:pyls (plugins :jedi_completion (:include_params t))))))
json-encode:    {\"python-mode\":{\"eglot-workspace-configuration\":{\"pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}}
json-serialize: {\"python-mode\":{\"eglot-workspace-configuration\":{\":pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}}"

"((python-mode (eglot-workspace-configuration (:pyls plugins (:jedi_completion (:include_params t))))))
json-encode:    {\"python-mode\":{\"eglot-workspace-configuration\":{\"pyls\":[\"plugins\",{\"jedi_completion\":{\"include_params\":true}}]}}}
json-serialize: nil"

"((python-mode (eglot-workspace-configuration (:pyls (:plugins :jedi_completion (:include_params t))))))
json-encode:    {\"python-mode\":{\"eglot-workspace-configuration\":{\"pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}}
json-serialize: {\"python-mode\":{\"eglot-workspace-configuration\":{\":pyls\":{\":plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}}"

"((python-mode (eglot-workspace-configuration (:pyls (plugins :jedi_completion (:include_params t))))))
json-encode:    {\"python-mode\":{\"eglot-workspace-configuration\":{\"pyls\":{\"plugins\":{\"jedi_completion\":{\"include_params\":true}}}}}}
json-serialize: nil"
```
</details>

<details>
  <summary>test.el</summary>
  
```
(require 'eglot)
(require 'jsonrpc)

(defun test-toggle-jsonrpc (toggle)
  (defalias 'jsonrpc--json-encode
    (if toggle ;; (fboundp 'json-serialize)
        (lambda (object)
          (json-serialize object
                          :false-object :json-false
                          :null-object nil))
      (require 'json)
      (defvar json-false)
      (defvar json-null)
      (declare-function json-encode "json" (object))
      (lambda (object)
        (let ((json-false :json-false)
              (json-null nil))
          (json-encode object))))
    "Encode OBJECT into a JSON string."))


(defun eglot-dump-it (conf toggle)
  (test-toggle-jsonrpc toggle)
  (ignore-errors
    (jsonrpc--json-encode
     (cl-loop for (section . v) in conf
              collect (if (keywordp section)
                          section
                        (intern (format ":%s" section)))
              collect v))))


(dolist (x (list
            '((python-mode
               . ((eglot-workspace-configuration
                   . ((:pyls . (:plugins (:jedi_completion (:include_params t)))))))))
            '((python-mode
               . ((eglot-workspace-configuration
                   . ((:pyls . ((plugins . (:jedi_completion (:include_params t))))))))))
            '((python-mode
               . ((eglot-workspace-configuration
                   . ((:pyls . ("plugins" (:jedi_completion (:include_params t)))))))))
            '((python-mode
               . ((eglot-workspace-configuration
                   . ((:pyls . ((:plugins . (:jedi_completion (:include_params t))))))))))
            '((python-mode
               . ((eglot-workspace-configuration
                   . ((:pyls . (("plugins" . (:jedi_completion (:include_params t))))))))))))
  (let ((r1 (eglot-dump-it x nil))
        (r2 (eglot-dump-it x t)))
    (unless (equal  r1 r2)
      (print (format "%s
json-encode:    %s
json-serialize: %s" x r1 r2)))))
```
</details>

fbergroth commented:
I rebased, splitted commits, and added some tests. Testing is ugly though, since I can't reach into `jsonrpc--json-encode`.
With this, you should see what currently does not work (in my parent comment), and what works in this PR.

There's probably a bug now here
https://github.com/joaotavora/eglot/blob/06e6dd6693123a737fb883dcf277432d63d6518e/eglot.el#L2248
since alists are rewritten to be plists. I think we want to use the sanitized configuration, so we should either dump as an alist or change the logic in `eglot-handle-request`.

No docs yet. Given the length of this PR, it seems things are likely to change...

joaotavora commented:
I've cherry-picked one of these commits (the `eglot-show-workspace-configuration` one) and pushed it to master.  Thanks @fbergroth: it's really a good aid.  

I've also finally re-studied the issue so now it's in my mental cache (to be evicted in ~2 weeks I'd say).  As a consequence I overhauled the `README.md` considerably again (also in preparation for a proper Texinfo manual on this).  Please have a read and tell me if it's clearer and/or if bugs are found.  You can probably add examples of how to configure your favourite servers. 

One thing I don't like is how the examples still refer to `pyls`, but as I understand it, it's not the preferred Python LSP server in Python land (`pylsp` would be it).


The way I see it, `eglot-workspace-configuration` is no work of art, but it's not _that_ confusing either.  The top-level structure is an alist.  And the lower structures can be anything that `json-serialize` can grok, though keyword-value plists are recommended. 

We _could_, rather easily also support `plist`s as top-level structure.  This is surely @astoff's suggestion.  We could also use that `eglot-sanitize-configuration` to do some extra checks, but frankly I'd say `eglot-show-configuration` as a debuging aid is good enough.  Then again I don't use `eglot-workspace-configuraiton` myself in any project, so I'll keep this PR open.

doolio commented:
> I overhauled the README.md

It is much clearer now thank you.

> One thing I don't like is how the examples still refer to pyls, but as I understand it, it's not the preferred Python LSP server in Python land (pylsp would be it).
> You can probably add examples of how to configure your favourite servers.

Yes, pylsp is now favoured over pyls which I believe is no longer actively maintained. It would just be a matter of changing pyls to pylsp everywhere in your example. For what it is worth I have created a `.dir-locals.el` for all the available settings for the pylsp server. It already follows the form in your example. In doing so I found some discrepancies in the pylsp schema.json and have submitted a PR to correct them. Once approved I can add this `.dir-locals.el` here to the 'Show and Tell` section of the Discussions page. It may encourage others to do the same for other servers.

astoff commented:
I currently get this error when I try to call `eglot-show-workspace-configuration`
```
eglot-show-workspace-configuration: Cannot open load file: No such file or directory, json-mode
```

fbergroth commented:
> I get this error

Ah crap, I forgot that json-mode is not built in.

Also @joaotavora,

> The way I see it, `eglot-workspace-configuration` is no work of art, but it's not _that_ confusing either. The top-level structure is an alist. And the lower structures can be anything that `json-serialize` can grok, though keyword-value plists are recommended.

Yes, but it can get tricky if you want to use the same configuration in emacs compiled without libjansson, where `json-serialize` is not available. Then "anything that `json-serialize` can grok" can serialize differently.

I think you got a bug in the first example in the readme. For me, it serializes to `{":pyls": { ...`

astoff commented:
> The top-level structure is an alist. And the lower structures can be anything that json-serialize can grok, though keyword-value plists are recommended.

Right, and this special requirement on the top level can only add to confusion, right?

I added some specific comments to the new README directly to in the commit.

Here's one a more "global" comment: The entire ["Format"](https://github.com/joaotavora/eglot#format) section should not be needed. Instead, Eglot should be such that the following is all you need to say:

> `eglot-workspace-configuration` can have any value that `json-serialize` understands. Typically, it will be a nested structure of plists or alists, which are then sent to the server as a nested JSON object. The JSON true, false and null values are represented by the Lisp values `t` and `:json-false` and `nil` respectively. JSON arrays are represented by Lisp vectors.

joaotavora commented:
> I think you got a bug in the first example in the readme. For me, it serializes to {":pyls": { ...

That's if you call `json-serialize` on the whole object directly.  But why would you do that? You let Eglot serialize it.  You can simulate this with `M-x eglot-show-configuration`.

> Yes, but it can get tricky if you want to use the same configuration in emacs compiled without libjansson, where json-serialize is not available. Then "anything that json-serialize can grok" can serialize differently.

This is why plists are recommended.  As far as I understand those pitfalls are only when you put alists in the sub-section structure.

fbergroth commented:
> That's if you call `json-serialize` on the whole object directly. But why would you do that? You let Eglot serialize it. You can simulate this with `M-x eglot-show-configuration`.

Sorry, my mistake, I had some parts of this PR evaled.

joaotavora commented:
> Here's one a more "global" comment: The entire ["Format"](https://github.com/joaotavora/eglot#format) section should not be needed. Instead, Eglot should be such that the following is all you need to say:
> eglot-workspace-configuration can have any value that json-serialize understands. 

If we allow `eglot-workspace-configuration` to be a plist of anything -- as you suggest in #1033 -- then yes, that statement is mostly true but still has the caveats pointed out by @fbergroth. However:
* I'm not convinced the complexity is worth it 
*  #1033 isn't finished yet (it doesn't handle the `workspace/configuration` request).

joaotavora commented:
@doolio 

> For what it is worth I have created a .dir-locals.el for all the available settings for the pylsp server. It already follows the form in your example. In doing so I found some discrepancies in the pylsp schema.json and have submitted a PR to correct them. Once approved I can add this .dir-locals.el here to the 'Show and Tell` section of the Discussions page. It may encourage others to do the same for other servers.

Yes that would be welcome.

> Yes, pylsp is now favoured over pyls which I believe is no longer actively maintained. It would just be a matter of changing pyls to pylsp everywhere in your example. 

Really? Do they accept the same configuration settings exactly?

astoff commented:
> This is why plists are recommended. As far as I understand those pitfalls are only when you put alists in the sub-section structure.

By the way, there's no good reason to use keyword keys in those plists, and if we don't then we protect the user from even knowing about those pitfalls. So, concretely, say this in the [simple example](https://github.com/joaotavora/eglot#simple-example):
```
((python-mode
  . ((eglot-workspace-configuration
      .
      ;; the value in the format described above starts here
      ((pyls . (plugins (jedi_completion (include_params t
                                          fuzzy t)
                         pylint (enabled :json-false)))))
      ;; and ends here
      ))))
```

> I'm not convinced the complexity is worth it

I don't understand what you mean here: what is more complex?

> https://github.com/joaotavora/eglot/pull/1033 isn't finished yet (it doesn't handle the workspace/configuration request).

Do you agree `workspace/configuration` support is broken, or at least not completely conformant to the spec? If so, then we can either merge #1033 now and fix everything later, or fix first `workspace/configuration` first and then I'll adapt my PR.

> Really? Do they accept the same configuration settings exactly?

Yes, or at least for the most part. Just need to change the top level key from `pyls` to `pylsp`.

doolio commented:
> Really? Do they accept the same configuration settings exactly?

Yes, from what I can tell you simply need to change the top level key to `pylsp` from `pyls`. Saying that, if you compare [this](https://github.com/python-lsp/python-lsp-server/blob/develop/CONFIGURATION.md) (from pylsp) to [this](https://github.com/palantir/python-language-server/blob/develop/vscode-client/package.json) (from pyls) it seems more settings have since been added to pylsp but that is perhaps unsurprising if pyls is no longer being actively developed.

joaotavora commented:
@fbergroth #1033 has been merged (sort of), meaning that `eglot-workspace-configuration` can now be a plist fully.  The existing function `eglot--workspace-configuraiton-plist` (which could be renamed) would be a good place to put the existing "sanitization" logic.  I think it's decently useful to have this sanitizer. I'll keep the PR open in case you want to work on that.  Otherwise, I thank you for the work so far and you may close it (again).

fbergroth commented:
Thanks. Perhaps it's more useful to just do normalization rather than validation (to handle dotted configuration section lookups). I'll close this again for now.
